rust/src/core/common/gates.rs

```diff
@@ -1,477 +1,736 @@
+pub enum GateType {
+    /// Identity.
+    I,
+    /// Pauli-X.
+    X,
+    /// Pauli-Y.
+    Y,
+    /// Pauli-Z.
+    Z,
+    /// Hadamard.
+    H,
+    /// Phase.
+    S,
+    /// S† (conjugate transpose of S).
+    SDAG,
+    /// T.
+    T,
+    /// T† (conjugate transpose of T).
+    TDAG,
+    /// 90 degree rotation around X-axis.
+    RX90,
+    /// minus 90 degree rotation around X-axis.
+    RXM90,
+    /// 180 degree rotation around X-axis.
+    RX180,
+    /// 90 degree rotation around Y-axis.
+    RY90,
+    /// minus 90 degree rotation around Y-axis.
+    RYM90,
+    /// 180 degree rotation around Y-axis.
+    RY180,
+    /// 90 degree rotation around Y-axis.
+    RZ90,
+    /// minus 90 degree rotation around Z-axis.
+    RZM90,
+    /// 180 degree rotation around Z-axis.
+    RZ180,
+    /// Arbitrary rotation around X-axis.
+    RX,
+    /// Arbitrary rotation around Y-axis.
+    RY,
+    /// Arbitrary rotation around Z-axis.
+    RZ,
+    /// Arbitrary rotation around Z-axis, global phase chosen such that it
+    /// works as a submatrix for controlled phase operations.
+    Phase,
+    /// Same as Phase, but with θ = π/2^k​.
+    PhaseK,
+    /// Arbitrary rotation around X-, Y- and Z-axis.
+    R,
+    /// Swap.
+    SWAP,
+    /// Square root of Swap.
+    SQSWAP,
+    /// Abstract unitary gate with number of target qubits specified.
+    U(usize),
 }
 
-/// Enumeration of predefined quantum gates.
-pub enum Gates {
+/// An unbound gate type.
+///
+/// The variants in this type specify all parameters to determine the behaviour
+/// of the gate. However, these variants carry no information about the target
+/// qubits.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum UnboundGate<'matrix> {
+    /// Identity.
+    I,
+    /// Pauli-X.
+    X,
+    /// Pauli-Y.
+    Y,
+    /// Pauli-Z.
+    Z,
+    /// Hadamard.
+    H,
+    /// Phase.
+    S,
+    /// S† (conjugate transpose of S).
+    SDAG,
+    /// T.
+    T,
+    /// T† (conjugate transpose of T).
+    TDAG,
+    /// 90 degree rotation around X-axis.
+    RX90,
+    /// minus 90 degree rotation around X-axis.
+    RXM90,
+    /// 180 degree rotation around X-axis.
+    RX180,
+    /// 90 degree rotation around Y-axis.
+    RY90,
+    /// minus 90 degree rotation around Y-axis.
+    RYM90,
+    /// 180 degree rotation around Y-axis.
+    RY180,
+    /// 90 degree rotation around Y-axis.
+    RZ90,
+    /// minus 90 degree rotation around Z-axis.
+    RZM90,
+    /// 180 degree rotation around Z-axis.
+    RZ180,
+    /// Arbitrary rotation around X-axis with specified angle (θ).
+    RX(f64),
+    /// Arbitrary rotation around Y-axis with specified angle (θ).
+    RY(f64),
+    /// Arbitrary rotation around Z-axis with specified angle (θ).
+    RZ(f64),
+    /// Arbitrary rotation around Z-axis, global phase chosen such that it
+    /// works as a submatrix for controlled phase operations.
+    Phase(f64),
+    /// Same as Phase, but with θ = π/2^k​.
+    PhaseK(u64),
```

Comment:
Why don't we use `u32` here?

Reply:
I just used u64 to be consistent in the size with the doubles everywhere else. You could just as well argue that a u16 is enough as doubles have only 11 exponent bits. It's arbitrary.